src/VecSim/algorithms/brute_force/brute_force.h

```diff
@@ -83,3 +84,457 @@ class BruteForceIndex : public VecSimIndex {
+        this->vectorBlocks.pop_back();
+
+        // Resize and align the id2labelmapping.
+        size_t id2label_size = idToLabelMapping.size();
+        // If the new size is smaller by at least one block comparing to the id2labemapping
+        // align to be a multlipication of blocksize  and resize by one block.
+        if (count + vectorBlockSize <= id2label_size) {
+            size_t vector_to_align_count = id2label_size % vectorBlockSize;
+            this->idToLabelMapping.resize(id2label_size - vectorBlockSize - vector_to_align_count);
+        }
+    }
+
+    return true;
+}
+
+template <typename DataType, typename DistFuncType>
+double BruteForceIndex<DataType, DistFuncType>::getDistanceFrom(size_t label,
+                                                                const void *vector_data) {
+    auto optionalId = this->labelToIdLookup.find(label);
+    if (optionalId == this->labelToIdLookup.end()) {
+        return INVALID_SCORE;
+    }
+    idType id = optionalId->second;
+
+    // Get the vectorBlock and the relative index of the required id.
+    VectorBlock *req_vectorBlock = getVectorVectorBlock(id);
+    size_t req_rel_idx = getVectorRelativeIndex(id);
+
+    return this->dist_func(req_vectorBlock->getVector(req_rel_idx), vector_data, &this->dim);
+}
+
+template <typename DataType, typename DistFuncType>
+size_t BruteForceIndex<DataType, DistFuncType>::indexSize() const {
+    return this->count;
+}
+
+// Compute the score for every vector in the block by using the given distance function.
+template <typename DataType, typename DistFuncType>
+vecsim_stl::vector<DistFuncType> BruteForceIndex<DataType, DistFuncType>::computeBlockScores(
+    VectorBlock *block, const void *queryBlob, void *timeoutCtx, VecSimQueryResult_Code *rc) const {
+    size_t len = block->getLength();
+    vecsim_stl::vector<DistFuncType> scores(len, this->allocator);
+    for (size_t i = 0; i < len; i++) {
+        if (__builtin_expect(VecSimIndex::timeoutCallback(timeoutCtx), 0)) {
+            *rc = VecSim_QueryResult_TimedOut;
+            return scores;
+        }
+        scores[i] = this->dist_func(block->getVector(i), queryBlob, &this->dim);
+    }
+    *rc = VecSim_QueryResult_OK;
+    return scores;
+}
+
+template <typename DataType, typename DistFuncType>
+VecSimQueryResult_List
+BruteForceIndex<DataType, DistFuncType>::topKQuery(const void *queryBlob, size_t k,
+                                                   VecSimQueryParams *queryParams) {
+
+    VecSimQueryResult_List rl = {0};
+    void *timeoutCtx = queryParams ? queryParams->timeoutCtx : NULL;
+
+    this->last_mode = STANDARD_KNN;
+    DataType normalized_blob[this->dim]; // This will be use only if metric == VecSimMetric_Cosine.
+    if (this->metric == VecSimMetric_Cosine) {
+        // TODO: need more generic
+        memcpy(normalized_blob, queryBlob, this->dim * sizeof(DataType));
+        float_vector_normalize(normalized_blob, this->dim);
```

Comment:
Same here (is making it generic should be part of this PR? + avoid allocating on stack if not have to)